src/agent/block_store_services/block_store_fs.js

```diff
@@ -58,171 +60,178 @@ class BlockStoreFs extends BlockStoreBase {
-                    // but just in case something changes we perform it once again here explicitly
-                    if (err.code !== 'ENOENT') {
-                        failed_to_delete_block_ids.push(block_id);
-                    }
-                    // TODO handle failed deletions - report to server and reclaim later
-                    dbg.warn('delete block failed due to', err);
-                })
-            )
-            .then(() => ({
-                failed_block_ids: failed_to_delete_block_ids,
-                succeeded_block_ids: _.difference(block_ids, failed_to_delete_block_ids)
-            }));
+    async _delete_blocks(block_ids) {
+        const succeeded_block_ids = [];
+        const failed_block_ids = [];
+        await P.map_with_concurrency(10, block_ids, async block_id => {
+            try {
+                await this._delete_block(block_id);
+                succeeded_block_ids.push(block_id);
+            } catch (err) {
+                // treat ENOENT as success - 
+                // this check is already performed inside _delete_block by calling ignore_not_found
+                // but just in case something changes we perform it once again here explicitly
+                if (err.code === 'ENOENT') {
+                    succeeded_block_ids.push(block_id);
+                } else {
+                    failed_block_ids.push(block_id);
+                }
+                dbg.warn(`delete block ${block_id} failed due to`, err);
+            }
+        });
+        return { failed_block_ids, succeeded_block_ids };
     }
 
-    _delete_block(block_id) {
+    async _delete_block(block_id) {
+        const fs_context = this.fs_context;
         const block_path = this._get_block_data_path(block_id);
         const meta_path = this._get_block_meta_path(block_id);
-        let del_stat;
-        let md_del_stat;
         dbg.log1("delete block", block_id);
-        return Promise.all([
-                fs.promises.stat(block_path)
-                .catch(ignore_not_found)
-                .then(stat => {
-                    del_stat = stat;
-                    return fs.promises.unlink(block_path).catch(ignore_not_found);
-                }),
-                fs.promises.stat(meta_path)
-                .catch(ignore_not_found)
-                .then(stat => {
-                    md_del_stat = stat;
-                    return fs.promises.unlink(meta_path).catch(ignore_not_found);
-                })
-            ])
-            .then(() => {
-                if (this._usage && del_stat) {
-                    const usage = {
-                        size: -(del_stat.size + ((md_del_stat && md_del_stat.size) ? md_del_stat.size : 0)),
-                        count: -1
-                    };
-                    return this._update_usage(usage);
-                }
-            });
+
+        const [block_stat, meta_stat] = await Promise.all([
```

Comment:
old code guards meta_path deletion to be after block_path deletion, don't we want to keep this behavior in the new code?

Reply:
I don't think the old code did that - it was using Promise.all too and awaited those calls in parallel.

Reply:
right, I mistakenly thought that they are chained, resolving